---
title: "Data preprocessing and integration"
output: html_notebook
---


```{r, include=FALSE, eval=TRUE}
knitr::opts_chunk$set(warning=FALSE,message=FALSE)
```



```{r}
library(Seurat)
library(ggplot2)
library(tidyr)
library(dplyr)
library(scuttle)
library(scran)
library(ggpubr)
library(RColorBrewer)
getPalette = colorRampPalette(brewer.pal(9, "Set1"))
dir.create(file.path("figs","integrated_data"))
```

## load data and create Seurat object

```{r}
E85_pucks = c("201104_07","201104_08","201104_09","201104_12","201104_13","201104_14","201104_16","201104_17","201104_18","201104_19","201104_22","201104_23","201104_24","201104_26","201104_27","201104_28","201104_29")

srt_list = list()
for (sel_puck in E85_pucks) {
  print(sel_puck)
  pos = read.csv(paste0("E8_5_data/",sel_puck,"_matched_bead_locations.txt"),sep="\t",header = F)
  cnt = read.csv(paste0("E8_5_data/",sel_puck,".digital_expression.txt.gz"),sep="\t",row.names = 1)
  
  pos = pos[,2:3]
  rownames(pos) = colnames(cnt)
  colnames(pos) = c("xcoord","ycoord")
  srt = CreateSeuratObject(counts = cnt,meta.data = pos,min.features=200,project = sel_puck)
  srt$stage = "E8.5"
  crop_f = paste0("E8_5_data/cropped_locations_DGEs/",sel_puck,"_cropped_locations.csv")
  if(file.exists(crop_f)){
    crop_reg = read.csv(crop_f,sep=",",header = T,row.names = 1)
    srt = srt[,colnames(srt) %in% crop_reg$barcode]
  }else{
    print(paste0(sel_puck," no selected region file found"))
  }
  srt_list[[sel_puck]] = srt
}

E95_pucks = c("201104_32","201104_33","201104_36","201112_03","201112_04","201112_05")

for (sel_puck in E95_pucks) {
  print(sel_puck)
  pos = read.csv(paste0("E9_5_data/",sel_puck,"_matched_bead_locations.txt"),sep="\t",header = F)
  cnt = read.csv(paste0("E9_5_data/",sel_puck,".digital_expression.txt.gz"),sep="\t",row.names = 1)
  
  pos = pos[,2:3]
  rownames(pos) = colnames(cnt)
  colnames(pos) = c("xcoord","ycoord")
  srt = CreateSeuratObject(counts = cnt,meta.data = pos,min.features=200,project = sel_puck)
  srt$stage = "E9.5"
  srt_list[[sel_puck]] = srt
}


srt = merge(srt_list[[1]],y=srt_list[-1],add.cell.ids=names(srt_list))
emb = srt@meta.data[,c("xcoord","ycoord")]
colnames(emb) = c("s_1","s_2")
srt[["spatial"]] <- CreateDimReducObject(embeddings = as.matrix(emb), key = "s_")
srt[["percent.mt"]] <- PercentageFeatureSet(srt, pattern = "^mt-")

srt <- subset(srt, subset = percent.mt < 20)
```

```{r}
firstup <- function(x) {
  x <- tolower(x)
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}
```

```{r,echo=FALSE,warning=FALSE,message=FALSE}
srt <- NormalizeData(srt)
srt <- FindVariableFeatures(srt,nfeatures =3000)
srt <- ScaleData(srt)
srt <- RunPCA(srt, features = VariableFeatures(srt))

srt <- CellCycleScoring(srt, s.features =firstup( cc.genes$s.genes), g2m.features =firstup( cc.genes$g2m.genes), set.ident = F)
srt <- FindNeighbors(srt , dims = 1:40, nn.eps = 0.1,verbose = F)
srt <- FindClusters(srt, resolution = 1.4,n.start = 10,verbose = F)
srt <- RunUMAP(srt, dims = 1:40, min.dist = 0.1,verbose = F)
```


```{r}
min_count = 500  # minimal mean UMI count per cluster, remove clusters that have very low counts
```


```{r}
srt.marker = FindAllMarkers(srt,min.diff.pct=0.05,only.pos = T,verbose=F,max.cells.per.ident=3000)
top_marker <- srt.marker[srt.marker$p_val_adj<0.01,] %>% group_by(gene) %>% top_n(n=1,wt=avg_logFC) %>% group_by(cluster) %>% summarise(cnt=n())

cluster_no_marker = unique(srt$seurat_clusters)[!(unique(srt$seurat_clusters) %in% top_marker$cluster)]

avg_cnts = srt@meta.data %>% group_by(seurat_clusters) %>% summarise(m_cnt=mean(nCount_RNA))

cluster_no_cnt = avg_cnts$seurat_clusters[avg_cnts$m_cnt<min_count]

removed_cluster = intersect(cluster_no_cnt,cluster_no_marker)
removed_cluster
```


```{r,echo=FALSE,warning=FALSE,message=FALSE}
if (length(removed_cluster)>0){
  srt = subset(srt,idents=removed_cluster,invert=T)
  srt <- NormalizeData(srt)
  srt <- FindVariableFeatures(srt,nfeatures =3000)
  srt <- ScaleData(srt)
  srt <- RunPCA(srt, features = VariableFeatures(srt))

  srt <- FindNeighbors(srt , dims = 1:40, nn.eps = 0.1,verbose = F)
  srt <- FindClusters(srt, resolution = 1.4,n.start = 10,verbose = F)
  srt <- RunUMAP(srt, dims = 1:40, min.dist = 0.1,verbose = F)
}

```


```{r}
srt$log_nCount_RNA = log10(srt$nCount_RNA)
srt$log_nFeature_RNA = log10(srt$nFeature_RNA)
VlnPlot(srt,features = c("log_nCount_RNA","log_nFeature_RNA","percent.mt"),group.by = "orig.ident",ncol=1,pt.size = 0)
ggsave(file.path("figs","integrated_data","vln_qc_pucks.pdf"),width = 12,height = 8)

VlnPlot(srt,features = c("log_nCount_RNA","log_nFeature_RNA","percent.mt"),group.by = "stage",ncol=1,pt.size = 0)
ggsave(file.path("figs","integrated_data","vln_qc_stage.pdf"),width = 12,height = 8)

VlnPlot(srt,features = c("log_nCount_RNA","log_nFeature_RNA","percent.mt"),group.by = "seurat_clusters",ncol=1,pt.size = 0)
ggsave(file.path("figs","integrated_data","vln_qc_clusters.pdf"),width = 12,height = 8)
```

```{r}
p1=ggplot(data=srt@meta.data,aes(x=nCount_RNA,fill=stage))+
  geom_histogram(position = "identity",alpha=0.5,bins = 200)+
  scale_x_continuous(limits=c(1,max(srt$nCount_RNA)),trans="log10") +
  scale_fill_brewer(palette = "Dark2")+
  labs(x="UMIs per bead",y="Number of beads",fill="Stage")+
  theme_classic()
p2=ggplot(data=srt@meta.data,aes(x=nFeature_RNA,fill=stage))+
  geom_histogram(position = "identity",alpha=0.5,bins = 200)+
  scale_fill_brewer(palette = "Dark2")+
  #scale_x_continuous(limits=c(1,max(srt$nFeature_RNA)),trans="log10") +
  scale_x_log10() +
  labs(x="Number of genes detected per bead",y="Number of beads",fill="Stage")+
  theme_classic()
ggarrange(p1,p2,ncol=1,common.legend=T)
ggsave(file.path("figs","integrated_data","hist_cnt_feature_qc_bystage.pdf"),width = 4.5,height = 4)
```

```{r,fig.width=16,fig.height=14}
DimPlot(srt,label = T,split.by = "orig.ident",ncol=4)
ggsave(file.path("figs","integrated_data","umap_cluster_split.png"),dpi=500,width = 16,height = 14)
```

```{r,fig.width=16,fig.height=14}
DimPlot(srt,label = F,reduction = "spatial",split.by="orig.ident",pt.size=0.01,ncol=4)
ggsave(file.path("figs","integrated_data","spatial_cluster_split.png"),dpi=500,width = 20,height = 20)
```


```{r,fig.width=9,fig.height=12}
srt.marker = FindAllMarkers(srt,min.diff.pct=0.05,only.pos = T,verbose=F,max.cells.per.ident=1000)
top10 <- srt.marker %>% group_by(cluster) %>% top_n(n = 20, wt = -p_val_adj) # %>%  top_n(n = 5, wt = avg_logFC)
top10 = top10 %>% group_by(gene) %>% top_n(n=1,wt=avg_logFC)
top10 = top10 %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC)

DotPlot(srt,features = unique(top10$gene),cluster.idents=F,cols="Spectral")+coord_flip()
ggsave(file.path("figs","integrated_data","marker_genes_per_cluster.pdf"),width = 12,height = 24)
```

## cell state annotation using known reference.

The scRNAseq reference data comes from: http://www.nature.com/articles/s41586-020-2552-x

```{r}
load("/stornext/Genomics/data/CLL_venetoclax/data_analysis/slideseq/slideseq/mouse_embryo/WT_E85_Seurat.Robj")
WT = WT_E85
# load("/stornext/Genomics/data/CLL_venetoclax/data_analysis/slideseq/slideseq/mouse_embryo/WT.Robj")
# WT = WT[,WT$stage %in% c("WT_80","WT_85")]
# WT = WT[,WT$cluster %in% names(table(WT$cluster))[table(WT$cluster)>25]]
 WT = FindVariableFeatures(WT,nfeatures = 3000)
 WT = ScaleData(WT)
 WT = RunPCA(WT)
anchors <- FindTransferAnchors(reference = WT, query = srt, npcs = 40,
    dims = 1:40)
predictions <- TransferData(anchorset = anchors, refdata = WT$cluster, 
    dims = 1:40)

#srt$predicted.id = predictions$predicted.id
srt <- AddMetaData(srt, metadata = predictions)
```



```{r}
merge_meta = rbind(WT@meta.data[,c("nCount_RNA", "nFeature_RNA")],
                   srt@meta.data[,c("nCount_RNA", "nFeature_RNA")])
merge_meta$protocol="Slide-seq V2"
merge_meta$protocol[grepl("^WT",rownames(merge_meta))]="10X single cell"
p1 = ggplot(data=merge_meta,aes(x=nFeature_RNA,fill=protocol))+
  geom_density(alpha=0.3)+
  scale_fill_brewer(palette = "Set1")+
  scale_x_log10() +
  labs(x="Number of genes detected per bead",y="Density",fill="")+
  theme_classic()

p2 = ggplot(data=merge_meta,aes(x=nCount_RNA,fill=protocol))+
  geom_density(alpha=0.3)+
  scale_fill_brewer(palette = "Set1")+
  scale_x_log10() +
  labs(x="UMIs per bead",y="Density",fill="")+
  theme_classic()

ggarrange(p1,p2,ncol=1,common.legend=T)
ggsave(file.path("figs","integrated_data","hist_qc_compare210X.pdf"),width = 4.5,height = 2.5)


```


```{r}
p1 = DimPlot(srt,label = T,pt.size=0.01)+ggtitle("de+novo clustering")
p2 = DimPlot(srt,label = T,pt.size=0.01,group.by = "predicted.id")+ggtitle("prediction based on reference")
p1 + p2
ggsave(file=file.path("figs","integrated_data","umap_denovo_prediction.png"),width = 14,height = 6)
```


```{r,fig.width=16,fig.height=14}
DimPlot(srt,label = F,reduction = "spatial",group.by = "predicted.id",split.by="orig.ident",pt.size=0.01,ncol=4)
ggsave(file.path("figs","integrated_data","spatial_predicted_split.png"),dpi=500,width = 20,height = 20)
```

```{r}
DimPlot(srt[,srt$orig.ident=="201104_32" & srt$predicted.id %in% c(33,36,39)],label = F,reduction = "spatial",group.by = "predicted.id",pt.size=0.01)
```

```{r,eval=FALSE}
DimPlot(srt[,srt$orig.ident=="201104_32" & srt$predicted.id %in% c(33,36,39)],label = F,reduction = "spatial",group.by = "predicted.id",pt.size=0.01)

plot_df = srt@meta.data[srt$orig.ident=="201104_32" & srt$predicted.id %in% c(33,36,39),]
ggplot(data=plot_df,aes(x=xcoord,y=ycoord,col=predicted.id))+
  geom_point(size=0.5)+
  guides(col = guide_legend(override.aes = list(size = 3)))+
  theme_classic()
```

```{r}
cluster_corr = table(srt$predicted.id,srt$seurat_clusters)

pheatmap::pheatmap(cluster_corr,scale = "column",display_numbers=cluster_corr,
                   main="column scaled. row:de-novo clustering. column: prediction",
                   filename =file.path("figs","integrated_data","heatmap_denovoVSprediction_colscale.pdf"),width = 11.5,height = 9 )
pheatmap::pheatmap(cluster_corr,scale = "row",
                   main="row scaled. row:de-novo clustering. column: prediction",
                   filename =file.path("figs","integrated_data","heatmap_denovoVSprediction_rowscale.pdf"))

pheatmap::pheatmap(table(srt$seurat_clusters,srt$stage),display_numbers=T,main="cluster (de-novo res=1.4) distribution across stage.",file=file.path("figs","integrated_data","heatmap_denovo_bystage.pdf"),width = 3.5,height = 9)

pheatmap::pheatmap(table(srt$predicted.id,srt$stage),display_numbers=T,main="cluster (annotated) distribution across stage.",file=file.path("figs","integrated_data","heatmap_predicted_bystage.pdf"),width = 3.5,height = 9)
```




```{r}
#saveRDS(srt,file = "data/srt_merged.Rds")
saveRDS(srt,file="data/srt_all_merged.Rds")
```

## data integration with reference scRNAseq data

remove beads with low prediction score, which indicate mixed cell or low quality

```{r}
srt = srt[,srt$prediction.score.max>0.4 & srt$nCount_RNA>500]
```



---
title: "downstream process - all data"
output: html_notebook
---

```{r}
library(Seurat)
library(SeuratWrappers)
library(ggplot2)
library(tidyr)
library(dplyr)
#library(scuttle)
library(patchwork)
#library(scran)
library(ggpubr)
library(RColorBrewer)
library(harmony)
getPalette = colorRampPalette(brewer.pal(9, "Set1"))
fig_dir = file.path("figs","all_data")
dir.create(fig_dir)

cluster_colors <- c('black', '#277424', '#CBBBA0', '#92C464', '#954B97', '#BD6FAB', '#581D6F', '#3AAA35', '#662483', '#9E937E', '#30892B', '#00A19A', '#64358C', '#8763A1', '#66296B', '#4A712E',
                    '#FFD100', '#B6A8D3', '#F9B233', '#1A5B1A', '#783F91', '#8C1D82', '#1C4024', '#107F71', '#7C4978', '#FFED00', '#7A2182', '#009767', '#624758', '#DF76AC', '#B670AC', '#6B3064',
                    '#6C6556', '#5567AE', '#477EC0', '#642F2C', '#318F2D', '#5B57A2', '#536C87', '#E94E1B', '#2FAC66', '#E20146')
cluster_order <- c('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30',
                   '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41')
names(cluster_colors) = cluster_order
```

## load data

```{r}
srt = readRDS(file = "/home/rstudio/processed_data/srt_all_merged.Rds")


```




```{r,fig.width=15,fig.height=15}
DimPlot(srt,reduction="spatial",split.by = "orig.ident",ncol=5)
```

```{r}
srt = srt[,srt$prediction.score.max>0.4 & srt$nCount_RNA>500]
```


```{r,fig.width=15,fig.height=15}
DimPlot(srt,reduction="spatial",split.by = "orig.ident",ncol=5)
```


```{r}
load("~/scRNAseq_reference_datasets/WT.Robj")

srt = merge(srt,WT)
srt$batch = srt$stage
srt$batch[grepl("^WT",srt$batch)] = "Ref"
```







```{r,warning=FALSE,message=FALSE}
ifnb.list <- SplitObject(srt, split.by = "batch")
ifnb.list <- lapply(X = ifnb.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 3000)
})

features <- SelectIntegrationFeatures(object.list = ifnb.list)
ifnb.list <- lapply(X = ifnb.list, FUN = function(x) {
    x <- ScaleData(x, features = features, verbose = FALSE)
    x <- RunPCA(x, features = features, verbose = FALSE)
})

immune.anchors <- FindIntegrationAnchors(object.list = ifnb.list, anchor.features = features, reduction = "rpca")
srt <- IntegrateData(anchorset = immune.anchors)
rm(immune.anchors)
rm(ifnb.list)
gc()
```


```{r}
srt <- ScaleData(srt, verbose = FALSE)
srt <- RunPCA(srt, npcs = 40, verbose = FALSE)
srt <- RunUMAP(srt, reduction = "pca", dims = 1:40)
srt <- FindNeighbors(srt, reduction = "pca", dims = 1:40)
srt <- FindClusters(srt, resolution = 0.8)
```


```{r}
DimPlot(srt,label=T)
```


```{r,fig.width=16,fig.height=16}
DimPlot(srt,split.by = "stage",label=T,ncol=3,group.by = "cluster")
```


```{r}
DimPlot(srt, group.by = "predicted.id",order=T,label = T)
```

```{r}
DimPlot(srt, group.by = "stage",order=T,label = F)
```


```{r}
srt$combined_cluster =srt$cluster
srt$combined_cluster[is.na(srt$combined_cluster)] = srt$predicted.id[is.na(srt$combined_cluster)]
```

```{r}
DimPlot(srt, group.by = "combined_cluster",order=T,label = T,cols = cluster_colors)
```

```{r}
saveRDS(srt,file="data/srt_combined_seurat.Rds")
```

```{r}
srt = readRDS(file="data/srt_combined_seurat.Rds")
```


```{r}
DimPlot(srt, group.by = "combined_cluster",label = T,repel = T)
ggsave(file.path(fig_dir,"umap_int_combined_annotation.png"),dpi = 300)
```


```{r}
srt$combined_stage = srt$stage
srt$combined_stage[srt$combined_stage=="WT_65"] = "E6.5"
srt$combined_stage[srt$combined_stage=="WT_70"] = "E7.0"
srt$combined_stage[srt$combined_stage=="WT_75"] = "E7.5"
srt$combined_stage[srt$combined_stage=="WT_80"] = "E8.0"
srt$combined_stage[srt$combined_stage=="WT_85"] = "E8.5"
```

```{r}
plot_df = data.frame(x=srt@reductions$umap@cell.embeddings[,1],
                     y=srt@reductions$umap@cell.embeddings[,2],
                     stage=srt$combined_stage)

pp_list = list()

for (sel_st in c("E6.5","E7.0","E7.5","E8.0","E8.5","E9.5")) {
  plot_df$hi_stage= (plot_df$stage ==sel_st)
  plot_df = plot_df[order(plot_df$hi_stage),]
  pp_list[[sel_st]] = ggplot(data=plot_df,aes(x=x,y=y,col=hi_stage))+
    geom_point(size=0.1,alpha=0.6)+
    scale_color_manual(values = c("grey80","black"))+
    theme_void()+
    labs(title=sel_st)+
    theme(legend.position = "none")
}


ggarrange(plotlist = pp_list,ncol=3,nrow=2)
ggsave(filename = file.path(fig_dir,"umap_stage_split.png"),dpi = 500)
```


```{r}
plot_df = data.frame(x=srt@reductions$umap@cell.embeddings[,1],
                     y=srt@reductions$umap@cell.embeddings[,2],
                     cluster=srt$combined_cluster)

pp_list = list()

for (sel_st in unique(srt$combined_cluster)  ) {
  plot_df$hi_cluster= (plot_df$cluster ==sel_st)
  plot_df = plot_df[order(plot_df$hi_cluster),]
  pp_list[[sel_st]] = ggplot(data=plot_df,aes(x=x,y=y,col=hi_cluster))+
    geom_point(size=0.01,alpha=0.5)+
    scale_color_manual(values = c("grey50","black"))+
    theme_void()+
    labs(title=sel_st)+
    theme(legend.position = "none")
}


pp = ggarrange(plotlist = pp_list,ncol=6,nrow=7)
ggsave(filename = file.path(fig_dir,"umap_cluster_split.png"),plot = pp,dpi = 300,width = 12,height = 14)
```











